lib/sensors/GPS.h

```diff
@@ -0,0 +1,362 @@
+//Protocol Specification: https://www.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_%28UBX-13003221%29.pdf
+//
+
+/*
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+associated documentation files (the "Software"), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to
+do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all copies or substantial
+portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+#include <array>
+#include "mbed.h"
+#pragma once
+
+//NOTE: There are a LOT of constants that follow, most unused
+//but I'm keeping them for convenience until I'm absolutely sure of what features we need
+//I do not envy the person who typed up all these values...
+//TODO: put these constants in a different header (maybe?)
+
+//Refresh frequency for new position data, in Hz
+//TODO: determine what frequency we want
+const uint8_t NAV_FREQUENCY = 5;
+const int ACK_TIMEOUT_MS = 2000;
+
+const uint32_t BUFFER_SIZE = 100;
+
+// ubxPacket validity
+typedef enum {
+  UBLOX_PACKET_VALIDITY_NOT_VALID,
+  UBLOX_PACKET_VALIDITY_VALID,
+  UBLOX_PACKET_VALIDITY_NOT_DEFINED,
+  UBLOX_PACKET_NOTACKNOWLEDGED // This indicates that we received a NACK
+} ublox_packet_validity_e;
+
+// Identify which packet buffer is in use:
+// packetCfg (or a custom packet), packetAck or m_packetBuf
+typedef enum {
+  UBLOX_PACKET_PACKETCFG,
+  UBLOX_PACKET_PACKETACK,
+  UBLOX_PACKET_PACKETBUF
+} ublox_packet_buffer_e;
+
+//header constants (first 2 message bytes)
+const uint8_t UBX_SYNCH_1 = 0xB5; //mu (greek letter)
+const uint8_t UBX_SYNCH_2 = 0x62; //b
+
+//receiver default baud rate
+const uint32_t DEFAULT_BAUD_RATE = 9600;
+
+//used in config to disable the default messages
+const std::array<const uint8_t, 7> DEFAULT_ENABLED_NMEA = {
+		0x00,  //GxGGA (Global positioning system fix data)
+		0x01,  //GxGLL (latitude and long, whith time of position fix and status)
+		0x02,  //GxGSA (GNSS DOP and Active satellites)
+		0x03,  //GxGSV (GNSS satellites in view)
+		0x04,  //GxRMC (Recommended minimum data)
+		0x05,  //GxVTG (course over ground and Ground speed)
+		0x41  //GxTXT (text transmission)
+};
+
+//oh boy here we go...
+const int MAX_PAYLOAD_SIZE = 768;
+
+//message class constants
```

Comment:
are these used outside of this file? Will this be included in several files?

Give this a read:
https://www.fluentcpp.com/2019/07/23/how-to-define-a-global-constant-in-cpp/

Keep in mind const in headers is implicitly static in c++. This is because all things in namespace scope implicitly have internal linkage(read: static). This includes the global namespace.

Reply:
@jetkov Since moving to `std=c++2a` in the Gimbal PR, we should probs just adopt inling constants

Reply:
Interesting, I didn't know that all the constants would be re-declared. Based on your above comment I assume that I can just fix this using `inline`?

Reply:
the more compatible way is to make them static constexpr which implies inline